### PR TITLE
ddl, types: convert the binary default value to proper encoding

### LINKED ISSUE #30798: binary literal in gbk columns is not handled
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
create table t(a char(10) charset gbk default 0xd2bb);
```

### 2. What did you expect to see? (Required)

`no error`

### 3. What did you see instead (Required)

`ERROR 1067 (42000): Invalid default value for 'a'`

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #30798

Problem Summary:

The binary default value is not decoded the corresponding column charset.

### What is changed and how it works?

Add the conversion.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

**Review by Defined2014 — APPROVED** [review]

File: ddl/integration_test.go [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+		defVal string
+		result string
+	}{
+		{"char(10) charset gbk", "0xC4E3BAC3", "你好"},
+		{"varchar(10) charset gbk", "0xC4E3BAC3", "你好"},
+		{"char(10) charset utf8mb4", "0xE4BDA0E5A5BD", "你好"},
+		{"char(10) charset utf8mb4", "0b111001001011100010010110111001111001010110001100", "世界"},
+		{"bit(48)", "0xE4BDA0E5A5BD", "你好"},
+		{"enum('你好')", "0xE4BDA0E5A5BD", "你好"},
+		{"set('你好')", "0xE4BDA0E5A5BD", "你好"},
```
Comment (zimulala) [inline ddl/integration_test.go:41]: Please add test like `create table t(a char(10) charset gbk default '好');`

Comment (tangenta) [reply ddl/integration_test.go:42]: Updated.

File: types/datum.go [hunk 2]
```diff
@@ -222,6 +222,9 @@ func (d *Datum) GetStringWithCheck(sc *stmtctx.StatementContext, chs string) (st
 
 func findEncoding(sc *stmtctx.StatementContext, chs string) (enc charset.Encoding, skip bool) {
 	enc = charset.FindEncoding(chs)
+	if sc == nil {
```
Comment (zimulala) [inline types/datum.go:225]: When `sc` is nil?

Comment (tangenta) [reply types/datum.go:225]: https://github.com/pingcap/tidb/pull/31196/files#diff-b7c54f7dad7257cbe4fa04b8d76c9bed694e2e2e7c5599f45eb2172d8d4ff8a7R913

**Review by zimulala — APPROVED** [review]
LGTM

**Pull request merged**